src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -1494,36 +1494,39 @@
     (fs/delete csr-path)
     (log/debug (i18n/trs "Removed certificate request for {0} at ''{1}''" subject csr-path))))
 
-(schema/defn revoke-existing-cert!
-  "Revoke the subject's certificate. Note this does not destroy the certificate.
-   The certificate will remain in the signed directory despite being revoked."
+(schema/defn revoke-existing-certs!
+  "Revoke the subjects' certificates. Note this does not destroy the certificates.
+   The certificates will remain in the signed directory despite being revoked."
   [{:keys [signeddir cacert cacrl cakey infra-crl-path
            infra-node-serials-path enable-infra-crl]} :- CaSettings
-   subject :- schema/Str]
-  (let [serial  (-> (path-to-cert signeddir subject)
-                    (utils/pem->cert)
-                    (utils/get-serial))
+   subjects :- [schema/Str]]
+  (let [serials (map #(-> (path-to-cert signeddir %)
+                       (utils/pem->cert)
+                       (utils/get-serial))
+                  subjects)
         [our-full-crl & rest-of-full-chain] (utils/pem->crls cacrl)
-        new-full-crl (utils/revoke our-full-crl
+        new-full-crl (utils/revoke-multiple our-full-crl
                                    (utils/pem->private-key cakey)
                                    (.getPublicKey (utils/pem->ca-cert cacert cakey))
-                                   serial)
+                                   serials)
         new-full-chain (cons new-full-crl (vec rest-of-full-chain))]
     (write-crls new-full-chain cacrl)
-    (log/debug (i18n/trs "Revoked {0} certificate with serial {1}" subject serial))
 
     ;; Publish infra-crl if an infra node is getting revoked.
     (when (and enable-infra-crl
-               (fs/exists? infra-node-serials-path)
-               (seq-contains? (read-infra-nodes infra-node-serials-path) (str serial)))
-      (let [[our-infra-crl & rest-of-infra-chain] (utils/pem->crls infra-crl-path)
-            new-infra-crl (utils/revoke our-infra-crl
+               (fs/exists? infra-node-serials-path))
+      (let [infra-revocations (filter #(seq-contains?
```

Comment:
few of things.

We shouldn't put the `read-infra-nodes` - which reaches out to the file system and (should) always return the same data during this loop - in the fn that gets called for every serial to revoke. I'd pull that into its own binding above.

I would also probably rewrite this to use `set`, and `set/intersection` instead of `filter` & `contains`.

To keep with the original conditional above we should test that `infra-revocations` is not empty before continuing.

Reply:
Thanks I'll clean that up.